webhook/index.js

```diff
@@ -0,0 +1,71 @@
+var bunyan = require('bunyan');
+var request = require('superagent');
+var noCache = require('superagent-no-cache')
+
+var log = bunyan.createLogger({name: "webhook"});
+
+var RETRY_DELAY = 60;
+
+function assertMessageIsWebhookEvent(msg) {
+    return !!msg.event && !!msg.url && !!msg.createdAt;
+}
+
+require('amqplib/callback_api').connect(
+    'amqp://localhost',
+    (err, conn) => {
+        if (err != null) {
+            return log.error({error: err}, 'queue connection error');
+        }
+
+        conn.createChannel((err, ch) => {
+            if (err != null) {
+                return log.error({error: err}, 'queue channel creation error');
+            }
+
+            ch.assertQueue('webhooks');
+            ch.consume('webhooks', (msg) => {
+                if (msg !== null) {
+                    msg = JSON.parse(msg.content.toString());
+
+                    if (!assertMessageIsWebhookEvent(msg)) {
+                        // FIXME: major error here, should we stop the service ?
+                        log.error(
+                            { message: msg },
+                            'assertion failed: queue message is not a valid webhook'
+                        );
+                        return ;
+                    }
+
+                    log.info({ webhook: msg }, 'handling webhook');
+
+                    if (!!msg.lastTriedAt && (msg.lastTriedAt - msg.lastTriedAt > RETRY_DELAY)) {
+                        log.info({ webhook: msg }, 'delaying webhook');
+                        ch.nack(msg, false, true);
```

Comment:
> Still, the retry queue has a fixed delay. And one of the best practice about webhooks is to retry at increasing intervals. How could we make this work with "dead-letter exchange" ? 

As far as I understand, TTL is indeed set to the whole DLX (dead-letter exchange), not per message. However, it does not mean we have to call the hook each time a message is read.

Instead, we can increment a counter and send the message to the remote server only at certain values (each Fibonacci values, for example). The DLX only make sure we don't process a message more than once every x units of time (every minute, for example).

If we want the first tries to burst, we can still re-enqueue the message to the main ("process now") queue a few times until it is enqueued to the staging ("retry later") queue.

> Maybe we should do the dead-letter exchange implementation first and figure that out later.

Sure. Let's keep it simple for now.